### PR TITLE
Added a static method "drain" under JcTools with a generic consumer

### PR DESCRIPTION
**Pull request by mohitmahi**
# Changes 

* Remove unnecessary "queue.isEmpty()" check while adding each span to batch list.
* Use MpscArrayQueue drain util method, which handle all cases when queue is empty or size > maxExportBatchSize
* Added a static method "drain" under JcTools with a generic consumer to handle cases for any type of Queue or span. 
* This PR is basically improving testing coverage and minor code-efficiency  for feature added  through https://github.com/open-telemetry/opentelemetry-java/pull/2983  & https://github.com/open-telemetry/opentelemetry-java/pull/3034 

# Test/Benchmarking
* All existing Unit Test are passing (no regression)
* Added a new test class for JCTool.

#### Ran a local test with both flavor of Queue (``ArrayBlockingQueue`` &  ``MpscArrayQueue``)

-------
> Happy Case:  (Queue with N size containing N item, and maxExportBatchSize = N)  ✅

Step1: Added 10000000 (N) items to both queue.
Step2: Create two Runnable task to call ``drain`` method, each with one of the queue type.
Step3: Submitted both task to a fixed threadpool executor service.
Step4: Measured total duration to drain 10000000 (N) from each task.

> Result ::

### ``ArrayBlockingQueue`` Duration:: ``495 ms`` added 10000000 to batch.

-------
> 2nd Case:  (Queue with N+100 size containing N+100 item, and maxExportBatchSize = N)  ✅

> Result ::

``MpscArrayQueue``      Added N item to batch.
``ArrayBlockingQueue``  Added N item to batch.

-------
> 3rd Case:  (Queue with N size containing N-100 item, and maxExportBatchSize = N)  ✅

> Result ::

``MpscArrayQueue``      Added N-100 item to batch.
``ArrayBlockingQueue``  Added N-100 item to batch.

-------
> 4th Case:  (Queue with N size containing 0 item, and maxExportBatchSize = N )  ✅

> Result ::

``MpscArrayQueue``      Added 0 item to batch.
``ArrayBlockingQueue``  Added 0 item to batch.

### CODE REVIEW

File: sdk/trace/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessorBuilder.java
Comment (anuraaga) [inline sdk/trace/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessorBuilder.java:70]: We don't need a test for normal queue with batchspanprocessor since it's not really supported. Let's revert those bits and just add a test in JcTools for the static method itself

Comment (mohitmahi) [reply sdk/trace/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessorBuilder.java]: @anuraaga Added a new test class for JcTools covering all cases. 

Screen Shot 2022-07-03 at 12 17 03 AM

File: sdk/trace-shaded-deps/src/main/java/io/opentelemetry/sdk/trace/internal/JcTools.java
Comment (anuraaga) [inline sdk/trace-shaded-deps/src/main/java/io/opentelemetry/sdk/trace/internal/JcTools.java:53]: Let's remove span from this file, can just be `T` and `item`.

**Review by anuraaga — APPROVED** [review]
Thanks, change looks good. Can you sign the CLA?

**Comment by mohitmahi** [comment]
Thanks @anuraaga for approving, the codecov check is failing but the new JcToolsTest covers 100% coverage of JcTool. Any idea ?

File: sdk/trace-shaded-deps/src/main/java/io/opentelemetry/sdk/trace/internal/JcTools.java
Comment (mohitmahi) [reply sdk/trace-shaded-deps/src/main/java/io/opentelemetry/sdk/trace/internal/JcTools.java:53]: the naming is just to improve readability here, as this method is mainly used here to pass Span type data only. If its a generic module/lib than "T" or "itemT"  will be a better choice. What do you think ?

Comment (anuraaga) [reply sdk/trace-shaded-deps/src/main/java/io/opentelemetry/sdk/trace/internal/JcTools.java:53]: We'll want to refactor to use this for logs too, there isn't any logic specific to spans so it's more confusing to use the word here.

**Review by anuraaga — COMMENTED** [review]
I think the codecov report is affected by us not directly publishing this module. It's ok to not worry about it, it doesn't block merging

File: sdk/trace-shaded-deps/src/main/java/io/opentelemetry/sdk/trace/internal/JcTools.java
Comment (mohitmahi) [reply sdk/trace-shaded-deps/src/main/java/io/opentelemetry/sdk/trace/internal/JcTools.java:53]: Okay will update in next rev.

File: sdk/trace-shaded-deps/src/main/java/io/opentelemetry/sdk/trace/internal/JcTools.java
Comment (anuraaga) [inline sdk/trace-shaded-deps/src/main/java/io/opentelemetry/sdk/trace/internal/JcTools.java:66]:
```suggestion
    T item;
    while (polledCount++ < maxExportBatchSize && (item = queue.poll()) != null) {
      consumer.accept(item);
```

File: sdk/trace-shaded-deps/src/main/java/io/opentelemetry/sdk/trace/internal/JcTools.java
Comment (jack-berg) [inline sdk/trace-shaded-deps/src/main/java/io/opentelemetry/sdk/trace/internal/JcTools.java:53]: This is a general purpose method so lets rename `maxExportBatchSize` to `limit` to match the naming from `MessagePassingQueue#drain`. Same applies to `drainNonJcQueue`.

File: sdk/trace-shaded-deps/src/test/java/io/opentelemetry/sdk/trace/internal/JcToolsTest.java
Comment (jack-berg) [inline sdk/trace-shaded-deps/src/test/java/io/opentelemetry/sdk/trace/internal/JcToolsTest.java:86]: Thanks for adding the additional test coverage!

#nit: can we rename the test names to `{methodUnderTest}_{Scenario}`? I.e. `capacity_MpscQueue`.

**Review by jack-berg — APPROVED** [review]
Couple small comments but looks good. Thanks!

**Comment by mohitmahi** [comment]
Thanks @jack-berg Addressed your comment. Do we need additional approval to get it merged ?

**Comment by mohitmahi** [comment]
@jkwatson @jack-berg @anuraaga can we get this PR merged, as all comments are addressed.

**Comment by anuraaga** [comment]
Thanks!

**Pull request merged**
